internal/resources/iampolicy/scope_helper.go

```diff
@@ -0,0 +1,156 @@
+	organization
+	clusterGroup
+	cluster
+	workspace
+	namespace
+)
+
+func stringToScope(s string) scopeType {
+	switch s {
+	case organizationKey:
+		return organization
+	case clusterGroupKey:
+		return clusterGroup
+	case clusterKey:
+		return cluster
+	case workspaceKey:
+		return workspace
+	case namespaceKey:
+		return namespace
+	default:
+		return unknown
+	}
+}
+
+func (s scopeType) String() string {
+	return [...]string{
+		"unknown",
+		organizationKey,
+		clusterGroupKey,
+		clusterKey,
+		workspaceKey,
+		namespaceKey,
+	}[s]
+}
+
+func validateScope(ctx context.Context, diff *schema.ResourceDiff, i interface{}) error {
+	value, ok := diff.GetOk(scopeKey)
+	if !ok {
+		return fmt.Errorf("scope: %v is not valid: minimum one valid scope block is required", value)
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return fmt.Errorf("scope data: %v is not valid: minimum one valid scope block is required", data)
+	}
+
+	scopeData := data[0].(map[string]interface{})
+	scopesFound := make([]string, 0)
+	scopesAllowed := []string{organizationKey, clusterGroupKey, clusterKey, workspaceKey, namespaceKey}
+
+	if v, ok := scopeData[organizationKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, organizationKey)
+		}
+	}
+
+	if v, ok := scopeData[clusterGroupKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, clusterGroupKey)
+		}
+	}
+
+	if v, ok := scopeData[clusterKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, clusterKey)
+		}
+	}
+
+	if v, ok := scopeData[workspaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, workspaceKey)
+		}
+	}
+
+	if v, ok := scopeData[namespaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, namespaceKey)
+		}
+	}
+
+	if len(scopesFound) == 0 {
+		return fmt.Errorf("no valid scope type block found: minimum one valid scope type block is required among: %v", scopesAllowed)
+	} else if len(scopesFound) > 1 {
+		return fmt.Errorf("found scopes: %v are not valid: maximum one valid scope type block is allowed", scopesFound)
+	}
+
+	return nil
+}
+
+func getScopeType(d *schema.ResourceData) (scopeType, error) {
+	scopes := []scopeType{}
+
+	value, ok := d.GetOk(scopeKey)
+	if !ok {
+		return unknown, fmt.Errorf("no scope defined")
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return unknown, fmt.Errorf("no type defined in scope")
+	}
+
+	scopeData, _ := data[0].(map[string]interface{})
+
+	for key, val := range scopeData {
+		v := val.([]interface{})
+		if len(v) == 0 {
+			continue
+		}
+
+		scopes = append(scopes, stringToScope(key))
+	}
+
+	if len(scopes) != 1 {
+		return unknown, fmt.Errorf("none or more than one scope types are defined")
+	}
+
+	return scopes[0], nil
+}
```

Comment:
Once schema validation is in place, these checks are redundant. 

This function can be combined with `constructScopeFullName` and can be called `constructScope` or  `constructScopedFullName`